### DISCUSSION TITLE
Pool.Close blocks indefinitely when a panic happens during iteration of statement rows

### DISCUSSION THREAD

**Discussion by comunidadio** [zombiezen/go-sqlite#73] [Q&A] [answered]
Managed to reproduce the issue with a minimal testcase:

```go
package main

import (
	"context"

	"zombiezen.com/go/sqlite/sqlitex"
)

var dbpool *sqlitex.Pool

func init() {
	var err error
	dbpool, err = sqlitex.NewPool("file::memory:?mode=memory", sqlitex.PoolOptions{})
	if err != nil {
		panic(err)
	}
}

func fini() {
	fmt.Println("i will block now...")
	dbpool.Close() // this blocks indefinitely
}

func querySomething() {
	conn := dbpool.Get(context.Background())
	defer dbpool.Put(conn)

	stmt := conn.Prep("SELECT 1")
	stmt.Step()
	panic("some error happened while iterating results...")
}

func main() {
	defer fini()

	querySomething()
}
```

**Comment by zombiezen** [comment, marked as answer]
Thanks for the minimal reproducer. It seems that what happens here is that there is an additional panic during `dbpool.Put` because the connection has a pending statement. If you add a `defer stmt.Reset()` after `conn.Prep`, the program completes as expected.

**Reply by comunidadio** [reply]
Thanks @zombiezen - I confirm defer stmt.Reset() makes the issue go away.

The indefinite blocking behaviour of pool.Close is surprising though, I see the check and panic in pool.Put now but shouldn't the program just indeed panic there instead of somehow swallowing it?

**Reply by zombiezen** [reply]
I always have to double-check what happens when a `panic` occurs during another `panic` being handled. `dbpool.Put` should be `panic`-ing, but I haven't delved too deeply to figure out why that's not showing up.
